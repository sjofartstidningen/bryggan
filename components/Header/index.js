import React from 'react';
import PropTypes from 'prop-types';
import Logotype from '../Logotype';
import {
  HeaderContainer,
  LogotypeContainer,
  NavContainer,
  Nav,
  NavItem,
  ProfileContainer,
  ProfileName,
  ProfileImage,
} from './components';

const Header = ({ links, user }) => (
  <HeaderContainer>
    <LogotypeContainer>
      <Logotype />
    </LogotypeContainer>

    <NavContainer className="navigation">
      <Nav>
        {links.map(link => (
          <NavItem key={link.href} active={link.active}>
            {link.title}
          </NavItem>
        ))}
      </Nav>
    </NavContainer>

    <ProfileContainer>
      <ProfileName>{user.name}</ProfileName>
      <ProfileImage>{user.img && <img src={user.img} alt="" />}</ProfileImage>
    </ProfileContainer>
  </HeaderContainer>
);

Header.propTypes = {
  links: PropTypes.arrayOf(
    PropTypes.shape({
      href: PropTypes.string.isRequired,
      title: PropTypes.string.isRequired,
      active: PropTypes.bool,
    }),
  ).isRequired,
  user: PropTypes.shape({
    name: PropTypes.string.isRequired,
    img: PropTypes.string,
  }).isRequired,
};

export default Header;
