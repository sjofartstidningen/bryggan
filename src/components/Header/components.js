import styled from 'styled-components';
import { modularScale, stripUnit, lighten } from 'polished';
import { NavLink } from 'react-router-dom';
import Logotype from '../Logotype';
import Gravatar from '../Gravatar';
import { ax } from '../../styles';

const modularScaleRem = x => `${stripUnit(modularScale(x))}rem`;

const Logo = styled(Logotype)`
  width: auto;
  height: ${modularScale(3)};
  margin-right: ${modularScaleRem(0)};
`;

const SiteTitle = styled.h1`
  margin-top: 0;
  margin-bottom: 0;
  font-size: ${modularScale(1)};
  font-weight: ${ax('fontWeight.bold')};
  letter-spacing: 0.05em;
`;

const NavList = styled.ul`
  display: flex;
  flex-flow: row nowrap;
  margin: 0;
  padding: 0;
  list-style: none;
`;

const NavItem = styled.li`
  margin-right: ${modularScale(0)};

  &:last-child {
    margin-right: 0;
  }
`;

const NavItemLink = styled(NavLink)`
  color: ${p => lighten(0.5, ax('color.black')(p))};
  text-decoration: none;
  transition: color 0.1s ease-in-out;

  &.active,
  &:hover {
    color: ${ax('color.black')};
  }
`;

const ProfileSection = styled.div`
  display: flex;
  align-items: center;
  margin-left: auto;

  & > * {
    margin-right: ${modularScaleRem(-1)};
  }

  & > *:last-child {
    margin-right: 0;
  }
`;

const ProfileImageContainer = styled.div`
  position: relative;

  &::after {
    content: '';
    position: absolute;
    top: -3px;
    left: -3px;
    height: calc(100% + 6px);
    width: calc(100% + 6px);
    border: 1px solid ${ax('color.black')};
    border-radius: 100%;
  }
`;

const ProfileImage = styled(Gravatar)`
  display: block;
  width: ${modularScale(2)};
  height: ${modularScale(2)};
  border-radius: 100%;
  background-color: white;
`;

const ProfileName = styled.p`
  margin-top: 0;
  margin-bottom: 0;
  font-weight: ${ax('fontWeight.normal')};
`;

const ProfileSignOut = ProfileName.extend`
  font-size: ${modularScale(-1)};
  font-weight: ${ax('fontWeight.normal')};
`;

const ProfileSignOutLink = styled.button`
  border: none;
  font-size: 1em;
  font-family: ${ax('font.sansSerif')}
  text-decoration: none;
  color: ${p => lighten(0.5, ax('color.black')(p))};
  background-color: transparent;
  transition: color 0.1s ease-in-out;

  &:hover {
    color: ${ax('color.warning')};
  }
`;

export {
  Logo,
  SiteTitle,
  NavList,
  NavItem,
  NavItemLink,
  ProfileSection,
  ProfileImageContainer,
  ProfileImage,
  ProfileName,
  ProfileSignOut,
  ProfileSignOutLink,
};
