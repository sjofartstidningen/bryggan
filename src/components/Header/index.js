/* eslint-disable no-nested-ternary */

import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Header as Wrapper } from '../MainGrid';
import { SignOut } from '../Icon';
import {
  Logo,
  SiteTitle,
  NavList,
  NavItem,
  NavItemLink,
  ProfileSection,
  ProfileImageContainer,
  ProfileImage,
  ProfileName,
  ProfileSignOut,
  ProfileSignOutLink,
} from './components';

class Header extends Component {
  static propTypes = {
    user: PropTypes.shape({
      displayName: PropTypes.string,
      email: PropTypes.string.isRequired,
    }),
    onSignOut: PropTypes.func,
  };

  static defaultProps = {
    user: null,
    onSignOut: null,
  };

  handleSignOut = () => this.props.onSignOut();

  render() {
    const { user } = this.props;

    return (
      <Wrapper>
        <Logo />
        <SiteTitle>Bryggan</SiteTitle>

        <nav>
          <NavList>
            <NavItem>
              <NavItemLink to="/tidningen">Tidningen</NavItemLink>
            </NavItem>
            <NavItem>
              <NavItemLink to="/installningar">Inst√§llningar</NavItemLink>
            </NavItem>
          </NavList>
        </nav>

        {user != null && (
          <ProfileSection>
            <ProfileImageContainer>
              <ProfileImage email={user.email} alt="" width="28" height="28" />
            </ProfileImageContainer>
            <ProfileName>{user.displayName || user.email}</ProfileName>
            <ProfileSignOut>
              <ProfileSignOutLink onClick={this.handleSignOut}>
                Logga ut <SignOut baseline />
              </ProfileSignOutLink>
            </ProfileSignOut>
          </ProfileSection>
        )}
      </Wrapper>
    );
  }
}

export default Header;
