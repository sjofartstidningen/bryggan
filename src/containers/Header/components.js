import styled from 'styled-components';
import { modularScale } from 'polished';
import { NavLink as Link } from 'react-router-dom';
import transition from '../../styles/transitions';

export const HeaderContainer = styled.header`
  position: relative;
  display: flex;
  align-items: center;
  width: 100vw;
  padding: ${modularScale(0)};
  font-size: ${modularScale(0)};
  font-family: ${props => props.theme.font.serif};
  z-index: ${props => props.theme.zIndex.top};
`;

export const LogotypeContainer = styled(Link)`
  display: block;
  width: 40px;
  margin-right: ${modularScale(1)};
`;

export const NavContainer = styled.div`
  margin-right: ${modularScale(1)};
`;

export const Nav = styled.ul`
  margin: 0;
  padding: 0;
  list-style: none;
`;

export const NavItem = styled.li`
  display: inline-block;
  margin-right: ${modularScale(-1)};
  border-right: 1px solid ${props => props.theme.color.grey};
  padding-right: ${modularScale(-1)};
  color: ${props =>
    props.active ? props.theme.color.black : props.theme.color.grey};

  &:last-child {
    margin: 0;
    border: none;
    padding: 0;
  }
`;

export const NavLink = styled(Link)`
  color: ${props => props.theme.color.grey};
  text-decoration: none;
  ${transition('color')};

  &:hover {
    color: ${props => props.theme.color.black};
  }

  &.active {
    color: ${props => props.theme.color.black};
  }
`;
