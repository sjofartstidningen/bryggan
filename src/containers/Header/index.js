import React from 'react';
import PropTypes from 'prop-types';
import config from '../../config';
import Logotype from '../../components/Logotype';
import {
  HeaderContainer,
  LogotypeContainer,
  NavContainer,
  Nav,
  NavItem,
  NavLink,
  ProfileContainer,
  ProfileName,
  ProfileImage,
  ProfileMenu,
  ProfileMenuItem,
  ProfileMenuLink,
  ProfileMenuInfo,
} from './components';
import IsHovering from '../../components/IsHovering';
import Settings from '../../components/Icons/Settings';
import LogOut from '../../components/Icons/LogOut';

function Header({ user }) {
  return (
    <HeaderContainer>
      <LogotypeContainer to="/">
        <Logotype />
      </LogotypeContainer>

      <NavContainer>
        <Nav>
          {config.mainLinks.map(link => (
            <NavItem key={link.href}>
              <NavLink to={link.href} activeClassName="active">
                {link.title}
              </NavLink>
            </NavItem>
          ))}
        </Nav>
      </NavContainer>

      {user ? (
        <IsHovering style={{ marginLeft: 'auto' }}>
          {({ hovering }) => (
            <ProfileContainer>
              <ProfileName>{user.name}</ProfileName>

              <ProfileImage>
                {user.img && <img src={user.img} alt="" />}
              </ProfileImage>

              <ProfileMenu hover={hovering}>
                <ProfileMenuItem>
                  <ProfileMenuLink href="#">
                    <Settings baseline /> Inst√§llningar
                  </ProfileMenuLink>
                </ProfileMenuItem>

                <ProfileMenuItem>
                  <ProfileMenuLink href="/auth/sign-out" warning>
                    <LogOut baseline /> Logga ut
                  </ProfileMenuLink>
                </ProfileMenuItem>

                <ProfileMenuItem>
                  <ProfileMenuInfo
                    href={config.repoUrl}
                    rel="noopener"
                    target="_blank"
                  >
                    {config.name} v{config.version}
                  </ProfileMenuInfo>
                </ProfileMenuItem>
              </ProfileMenu>
            </ProfileContainer>
          )}
        </IsHovering>
      ) : (
        <ProfileContainer>
          <NavLink to="/auth/sign-in" activeClassName="active">
            Logga in
          </NavLink>
        </ProfileContainer>
      )}
    </HeaderContainer>
  );
}

Header.propTypes = {
  user: PropTypes.shape({
    name: PropTypes.string.isRequired,
    img: PropTypes.string,
  }),
};

Header.defaultProps = { user: null };

export default Header;
