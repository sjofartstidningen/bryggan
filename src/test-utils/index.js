/* eslint-disable import/no-extraneous-dependencies, react/prop-types */
import React from 'react';
import { ThemeProvider } from 'styled-components';
import { BrowserRouter } from 'react-router-dom';
import { theme } from '../styles';

const WithContexts = ({ children }) => (
  <BrowserRouter>
    <ThemeProvider theme={theme}>{children}</ThemeProvider>
  </BrowserRouter>
);

function setupIntersectionObserver() {
  const original = window.IntersectionObserver;

  class IntersectionObserver {
    constructor(callback) {
      this.callback = callback;
    }

    entries = [];

    observe = el => {
      this.entries.push(el);
    };

    unobserve = el => {
      this.entries = this.entries.filter(x => x !== el);
    };

    simulateIntersectionEvent = (data = {}) => {
      const els = this.entries.map(entry => ({
        target: entry,
        ...data,
      }));

      this.callback(els);
    };
  }

  window.IntersectionObserver = IntersectionObserver;

  return () => {
    window.IntersectionObserver = original;
  };
}

const sleep = (ms = 0) =>
  new Promise(resolve => {
    window.setTimeout(resolve, ms);
  });

export { WithContexts, setupIntersectionObserver, sleep };
