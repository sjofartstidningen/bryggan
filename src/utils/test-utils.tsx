import React from 'react';
import { render, RenderOptions } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import { theme } from '../styles/theme';
import { AuthProvider } from '../hooks/useAuth';

const Providers: React.FC = ({ children }) => {
  return (
    <React.Suspense fallback={<p>Loading</p>}>
      <ThemeProvider theme={theme}>
        <AuthProvider>{children}</AuthProvider>
      </ThemeProvider>
    </React.Suspense>
  );
};

const customRender = (ui: React.ReactElement, options?: RenderOptions) =>
  render(ui, {
    wrapper: ({ children }) => <Providers>{children}</Providers>,
    ...options,
  });

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render };
