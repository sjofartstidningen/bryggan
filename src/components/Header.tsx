import React, { useState, useEffect } from 'react';
import styled from 'styled-components';
import { Link } from 'react-router-dom';
import { useLazyQuery } from '@apollo/react-hooks';
import gql from 'graphql-tag';
import { spacing, font, size, color, tracking, layer } from '../styles/theme';
import { boxShadow } from '../styles/utils';
import { useAuthState } from '../hooks/use-auth';
import { useScreens } from '../hooks/use-theme';
import { SearchBox } from './SearchBox';
import { ProfileBox } from './ProfileBox';
import { ScreenMinWidth } from './ScreenSize';
import { SettingsMenu } from './SettingsMenu';
import { UserQuery } from '../types/graphql';

export const USER_QUERY = gql`
  query UserQuery {
    currentAccount {
      email
      name {
        displayName
      }
      profilePhotoUrl
    }
  }
`;

export const Header: React.FC = () => {
  const state = useAuthState();
  const screens = useScreens();
  const [menuIsOpen, setMenuIsOpen] = useState(false);
  const [fetchUser, { data, called }] = useLazyQuery<UserQuery>(USER_QUERY);

  useEffect(() => {
    if (state.matches('authenticated') && !called) fetchUser();
  }, [state, called, fetchUser]);

  return (
    <HeaderContainer>
      <Title>
        <Link to="/">Bryggan</Link>
      </Title>

      {state.matches('authenticated') && (
        <>
          <ScreenMinWidth width={Number.parseInt(screens.sm, 10)}>
            <SearchContainer>
              <SearchBox />
            </SearchContainer>
          </ScreenMinWidth>

          {data && (
            <React.Fragment>
              <ProfileContainer>
                <ProfileBox
                  label="Toggle context menu"
                  profilePhotoUrl={data.currentAccount.profilePhotoUrl}
                  onClick={() => setMenuIsOpen(!menuIsOpen)}
                  aria-controls="context-menu"
                  aria-expanded={menuIsOpen}
                />
              </ProfileContainer>

              <SettingsMenu
                id="context-menu"
                isOpen={menuIsOpen}
                profile={data.currentAccount}
                onClose={() => setMenuIsOpen(false)}
              />
            </React.Fragment>
          )}
        </>
      )}
    </HeaderContainer>
  );
};

const HeaderContainer = styled.header`
  display: flex;
  flex-flow: row nowrap;
  justify-content: flex-start;
  align-items: center;
  width: 100vw;
  height: ${spacing('16')};
  padding: 0 ${spacing('4')};
  background-color: ${color('shade')};
  z-index: ${layer('header')};

  ${boxShadow('md')};
`;

const Title = styled.h1`
  margin: ${spacing('0')};
  margin-right: ${spacing('4')};
  padding: ${spacing('0')};
  font-family: ${font('display')};
  font-size: ${size('lg')};
  text-transform: lowercase;
  letter-spacing: ${tracking('widest')};

  a {
    color: ${color('black')};
    text-decoration: none;
  }
`;

const SearchContainer = styled.div`
  display: block;
  flex: 1;
  max-width: calc(${spacing('64')} * 2.5);
  margin-right: ${spacing('4')};
`;

const ProfileContainer = styled.div`
  margin-left: auto;
`;
