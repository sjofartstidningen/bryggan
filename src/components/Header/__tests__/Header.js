import React from 'react';
import { shallow, mount } from 'enzyme';
import { WithContexts } from '../../../test-utils';
import Header from '../';

describe('Component.Header', () => {
  test('should render without user', () => {
    const wrapper = shallow(<Header />);
    expect(wrapper).toMatchSnapshot();
  });

  test('should render child component if user is defined', () => {
    const user = { email: 'adam@fransvilhelm.com' };
    const wrapper = shallow(<Header user={user} />);

    expect(wrapper).toMatchSnapshot();
  });

  test('should call onSignOut-prop when clicking sign out button', () => {
    const user = { email: 'adam@fransvilhelm.com' };
    const handleSignOut = jest.fn();
    const wrapper = mount(
      <WithContexts>
        <Header user={user} onSignOut={handleSignOut} />
      </WithContexts>,
    );

    const button = wrapper.find('button');
    button.simulate('click', { clicked: true });

    expect(handleSignOut).toHaveBeenCalled();
  });
});
