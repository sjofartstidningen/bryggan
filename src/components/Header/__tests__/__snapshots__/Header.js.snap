// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component.Header should render user component if user is defined 1`] = `
<styled.header>
  <Styled(Logotype) />
  <styled.h1>
    Bryggan
  </styled.h1>
  <nav>
    <styled.ul>
      <styled.li>
        <Styled(NavLink)
          to="/tidningen"
        >
          Tidningen
        </Styled(NavLink)>
      </styled.li>
      <styled.li>
        <Styled(NavLink)
          to="/installningar"
        >
          Inställningar
        </Styled(NavLink)>
      </styled.li>
    </styled.ul>
  </nav>
  <styled.div>
    <styled.div>
      <Styled(Gravatar)
        alt=""
        email="adam@fransvilhelm.com"
        height="28"
        width="28"
      />
    </styled.div>
    <styled.p>
      adam@fransvilhelm.com
    </styled.p>
    <styled.p>
      <styled.button
        onClick={[Function]}
      >
        Logga ut 
        <SignOut
          baseline={true}
        />
      </styled.button>
    </styled.p>
  </styled.div>
</styled.header>
`;

exports[`Component.Header should render without user 1`] = `
<styled.header>
  <Styled(Logotype) />
  <styled.h1>
    Bryggan
  </styled.h1>
  <nav>
    <styled.ul>
      <styled.li>
        <Styled(NavLink)
          to="/tidningen"
        >
          Tidningen
        </Styled(NavLink)>
      </styled.li>
      <styled.li>
        <Styled(NavLink)
          to="/installningar"
        >
          Inställningar
        </Styled(NavLink)>
      </styled.li>
    </styled.ul>
  </nav>
</styled.header>
`;
